Add concepts page (#288)
ros-planning/moveit2_tutorials: doc/concepts/concepts.rst

========
Concepts
========

The following list major MoveIt concepts and components and describes how they tie in with each other. For tutorials, see the `tutorials <https://moveit.picknik.ai/>`_ page.

.. image:: /_static/images/moveit_pipeline.png

***********************
The ``move_group`` node
***********************

> The figure above shows the high-level system architecture for the primary node provided by MoveIt called `move_group <http://docs.ros.org/noetic/api/moveit_ros_move_group/html/move__group_8cpp.html>`_. This node serves as an integrator: pulling all the individual components together to provide a set of ROS actions and services for users to use.

Reviewer: ```suggestion
The figure below shows the high-level system architecture for a key node provided by MoveIt called `move_group <https://docs.ros.org/noetic/api/moveit_ros_move_group/html/move__group_8cpp.html>`_. This node serves as an integrator: pulling all the individual components together to provide a set of ROS actions and services for users to use. Note that moveit_commander is `not currently available in MoveIt 2 <https://github.com/ros-planning/moveit2/issues/314>`_.
```

moveit cpp probably didn't exist when this was written

Author: Yeah, I think moveitcpp didn't exist back then.

.. image:: /_static/images/move_group.png

User Interface
--------------

> The users can access the actions and services provided by ``move_group`` in three ways:

Reviewer: ```suggestion
The users can access the actions and services provided by ``move_group`` in two ways:
```

commander isn't available, we'll have to get around to updating diagram later

Configuration
-------------

move_group is a ROS node. It uses the ROS param server to get three kinds of information:

1. URDF - **move_group** looks for the ``robot_description`` parameter to get the URDF for the robot.

2. SRDF - **move_group** looks for the ``robot_description_semantic`` parameter to get the SRDF for the robot. The SRDF is typically created (once) by a user using the MoveIt Setup Assistant.

3. MoveIt configuration - ``move_group`` will look for other configuration specific to MoveIt including joint limits, kinematics, motion planning, perception and other information. Config files for these components are automatically generated by the MoveIt Setup Assistant and stored in the _config_ directory of the corresponding MoveIt config package for the robot.

Robot Interface
---------------

``move_group`` talks to the robot through ROS topics and actions. It communicates with the robot to get current state information (positions of the joints, etc.), to get point clouds and other sensor data from the robot sensors and to talk to the controllers on the robot.

Joint State Information
-----------------------

> ``move_group`` listens on the ``/joint_states`` topic for determining the current state information - i.e. determining where each joint of the robot is. ``move_group`` is capable of listening to multiple publishers on this topic even if they are publishing only partial information about the robot state (e.g. separate publishers may be used for the arm and mobile base of a robot). Note that ``move_group`` will not setup its own joint state publisher - this is something that has to be implemented on each robot.

Reviewer: ```suggestion
``move_group`` listens on the ``/joint_states`` topic for determining the current state information - i.e. determining where each joint of the robot is. ``move_group`` is capable of listening to multiple publishers on this topic even if they are publishing only partial information about the robot state (e.g. separate publishers may be used for the arm and mobile base of a robot). Note that ``move_group`` will not set up its own joint state publisher - this is something that has to be implemented on each robot.
```

nit - this typo is all over all our docs :(

AddTimeParameterization
^^^^^^^^^^^^^^^^^^^^^^^

The motion planners will typically generate "kinematic paths", i.e., paths that do not obey any velocity or acceleration constraints and are not time parameterized. This adapter will "time parameterize" the motion plans by applying velocity and acceleration constraints.

ResolveConstraintFrames
^^^^^^^^^^^^^^^^^^^^^^^

Goal constraints can be set using subframes (e.g. a pose goal in the frame ``cup/handle``, where ``handle`` is a subframe on the object ``cup``). This adapter changes the frame of constraints to an object or robot frame (e.g. ``cup``).

OMPL
----

OMPL (Open Motion Planning Library) is an open-source motion planning library that primarily implements randomized motion planners. MoveIt integrates directly with OMPL and uses the motion planners from that library as its primary/default set of planners. The planners in OMPL are abstract; i.e. OMPL has no concept of a robot. Instead, MoveIt configures OMPL and provides the back-end for OMPL to work with problems in Robotics.

.. raw:: html

   <hr>

**********************
Planning Scene Monitor
**********************

> The ``planning scene`` is an object used for storing the representation of the world around the robot and also the state of the robot itself. The internal state of the ``planning_scene`` object is typically maintained by ``planning_scene_monitor`` component that enables reading and writing the state in a thread-safe manner.

Reviewer: ```suggestion
The ``planning scene`` is an object used for storing the representation of the world around the robot and also the state of the robot itself. The internal state of the ``planning_scene`` object is typically maintained by a ``planning_scene_monitor`` component that enables reading and writing the state in a thread-safe manner.
```

nit - just running this file through a spell/grammar checker

.. image:: /_static/images/planning_scene_monitor.svg

World Geometry Monitor
----------------------

The world geometry monitor builds world geometry using information from the sensors on the robot such as LIDARs or depth cameras and from user input. It uses the ``occupancy map monitor`` described below to build a 3D representation of the environment around the robot and augments that with information on the ``planning_scene`` topic for adding object information.

3D Perception
-------------

3D perception in MoveIt is handled by the ``occupancy map monitor``. The occupancy map monitor uses a plugin architecture to handle different kinds of sensor input as shown in the Figure above. In particular, MoveIt has inbuilt support for handling two kinds of inputs:

- **Point clouds**: handled by the ``point cloud occupancy map updater`` plugin

- **Depth images**: handled by the ``depth image occupancy map updater`` plugin

Note that you can add your own types of updaters as a plugin to the occupancy map monitor.

Octomap
-------

> The Occupancy map monitor uses an `Octomap <http://octomap.github.io/>`_ to maintain the occupancy map of the environment. The Octomap can actually encode probabilistic information about individual cells although this information is not currently used in MoveIt. The Octomap can directly be passed into FCL, the collision checking library that MoveIt uses.

Reviewer: ```suggestion
The Occupancy map monitor uses an `Octomap <https://octomap.github.io/>`_ to maintain the occupancy map of the environment. The Octomap can actually encode probabilistic information about individual cells although this information is not currently used in MoveIt. The Octomap can directly be passed into FCL, the collision checking library that MoveIt uses.
```

Depth Image Occupancy Map Updater
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The depth image occupancy map updater includes its own *self-filter*, i.e. it will remove visible parts of the robot from the depth map. It uses current information about the robot (the robot state) to carry out this operation.

.. raw:: html

   <hr>

**********
Kinematics
**********

The Kinematics Plugin
---------------------

MoveIt uses a plugin infrastructure, especially targeted towards allowing users to write their own inverse kinematics algorithms. Forward kinematics and finding jacobians is integrated within the RobotState class itself. The default inverse kinematics plugin for MoveIt is configured using the `KDL <https://github.com/orocos/orocos_kinematics_dynamics>`_ numerical jacobian-based solver. This plugin is automatically configured by the MoveIt Setup Assistant.

******************
Collision Checking
******************

> Collision checking in MoveIt is configured inside a Planning Scene using the _CollisionWorld_ object. Fortunately, MoveIt is setup so that users never really have to worry about how collision checking is happening. Collision checking in MoveIt is mainly carried out using the `FCL <https://flexible-collision-library.github.io/>`_ package - the primary collision checking library of MoveIt.

Reviewer: ```suggestion
Collision checking in MoveIt is configured inside a Planning Scene using the _CollisionWorld_ object. Fortunately, MoveIt is set up so that users never really have to worry about how collision checking is happening. Collision checking in MoveIt is mainly carried out using the `FCL <https://flexible-collision-library.github.io/>`_ package - the primary collision checking library of MoveIt.
```

Collision Objects
-----------------

MoveIt supports collision checking for different types of objects including:

- **Meshes** - you can use either ``.stl`` (standard triangle language) or ``.dae`` (digital asset exchange) formats to describe objects such as robot links.

- **Primitive Shapes** - e.g. boxes, cylinders, cones, spheres and planes

- **Octomap** - the ``Octomap`` object can be directly used for collision checking

Allowed Collision Matrix (ACM)
------------------------------

> Collision checking is a very expensive operation often accounting for close to 90% of the computational expense during motion planning. The ``Allowed Collision Matrix`` or ``ACM`` encodes a binary value corresponding to the need to check for collision between pairs of bodies (which could be on the robot or in the world). If the value corresponding to two bodies is set to ``true`` in the ACM, it specifies that a collision check between the two bodies is either not required or wanted. The collision checking would not be required if, e.g., the two bodies are always so far way that they can never collide with each other. Alternatively, the two bodies could be in contact with each other by default, in which case the collision detection should be disabled for the pair in the ACM.

Reviewer: ```suggestion
Collision checking is a very expensive operation often accounting for close to 90% of the computational expense during motion planning. The ``Allowed Collision Matrix`` or ``ACM`` encodes a binary value corresponding to the need to check for collision between pairs of bodies (which could be on the robot or in the world). If the value corresponding to two bodies is set to ``true`` in the ACM, it specifies that a collision check between the two bodies is either not required or wanted. The collision checking would not be required if, e.g., the two bodies are always so far away that they can never collide with each other. Alternatively, the two bodies could be in contact with each other by default, in which case the collision detection should be disabled for the pair in the ACM.
```

.. raw:: html

   <hr>

*********************
Trajectory Processing
*********************

Time parameterization
---------------------

> Motion planners typically only generate “paths”, i.e. there is no timing information associated with the paths. MoveIt includes several `trajectory processing algorithms <http://docs.ros.org/en/noetic/api/moveit_core/html/cpp/classtrajectory__processing_1_1TimeOptimalTrajectoryGeneration.html>`_ that can work on these paths and generate trajectories that are properly time-parameterized accounting for the maximum velocity and acceleration limits imposed on individual joints. These limits are read from a special ``joint_limits.yaml`` configuration file that is specified for each robot. The configuration file is optional and it overrides any velocity or acceleration limits from the URDF. The recommended algorithm as of January 2022 is **time_optimal_trajectory_generation** (TOTG). A caveat for this algorithm is that the robot must start and end at rest. By default, the TOTG timestep is 0.1 seconds.

Reviewer: ```suggestion
Motion planners typically only generate “paths”, i.e. there is no timing information associated with the paths. MoveIt includes several `trajectory processing algorithms <https://docs.ros.org/en/noetic/api/moveit_core/html/cpp/classtrajectory__processing_1_1TimeOptimalTrajectoryGeneration.html>`_ that can work on these paths and generate trajectories that are properly time-parameterized accounting for the maximum velocity and acceleration limits imposed on individual joints. These limits are read from a special ``joint_limits.yaml`` configuration file that is specified for each robot. The configuration file is optional and it overrides any velocity or acceleration limits from the URDF. The recommended algorithm as of January 2022 is **time_optimal_trajectory_generation** (TOTG). A caveat for this algorithm is that the robot must start and end at rest. By default, the TOTG timestep is 0.1 seconds.
```

> ========

Reviewer 2: Would it be useful to break some of this up into separate sub-pages in the concepts section?  That'll make it easier to expand this in the future.

Author: Yeah, that makes sense to me.

Concepts
========

The following list major MoveIt concepts and components and describes how they tie in with each other. For tutorials, see the `tutorials <https://moveit.picknik.ai/>`_ page.

.. image:: /_static/images/moveit_pipeline.png

***********************
The ``move_group`` node
***********************

The figure above shows the high-level system architecture for the primary node provided by MoveIt called [move_group](http://docs.ros.org/noetic/api/moveit_ros_move_group/html/move__group_8cpp.html). This node serves as an integrator: pulling all the individual components together to provide a set of ROS actions and services for users to use.

.. image:: /_static/images/move_group.png

User Interface
--------------

The users can access the actions and services provided by ``move_group`` in three ways:

> - **In C++** - using the `move_group_interface <http://docs.ros.org/noetic/api/moveit_ros_planning_interface/html/classmoveit_1_1planning__interface_1_1MoveGroupInterface.html>`_ package that provides an easy to setup C++ interface to move_group

Reviewer 3: This link should be updated to point to rolling (or, even better, `${moveit_release}`, if Sphinx has some way to expand variables).

Author: I'll update this to point to the `rolling` API docs once they are online. There's nothing better than `noetic` available for now.

Reviewer 2: There is a way.  You use the variable `{DISTRO}`.  But as @jliukkonen said they don't exist yet so this is fine for now.

> - **In Python** - using the `moveit_commander <http://docs.ros.org/noetic/api/moveit_commander/html/classmoveit__commander_1_1move__group_1_1MoveGroupCommander.html>`_ package

Reviewer 3: moveit_commander doesn't exist for MoveIt 2. At the risk of stating the obvious, you could say that in Python (and Rust, Lisp, etc.,) you can interact with MoveIt at the message level.

Author: I think I'll just remove the Python entry then. :+1:

> - **Through a GUI** - using the `Motion Planning plugin to Rviz <https://moveit.picknik.ai/foxy/doc/quickstart_in_rviz/quickstart_in_rviz_tutorial.html>`_ (the ROS visualizer)

Reviewer 3: ```suggestion
- **Through a GUI** - using the :doc:`Motion Planning plugin to Rviz </doc/quickstart_in_rviz/quickstart_in_rviz_tutorial.html>` (the ROS visualizer)
```

Author: The suggested change doesn't seem to work... I'm trying to figure out how to do relative links in RST if you they are supported.

Author: Apparently like this: 
```doc:`Motion Planning plugin to Rviz <../tutorials/quickstart_in_rviz/quickstart_in_rviz_tutorial>` ```

Depth Image Occupancy Map Updater
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

The depth image occupancy map updater includes its own *self-filter*, i.e. it will remove visible parts of the robot from the depth map. It uses current information about the robot (the robot state) to carry out this operation.

**********
Kinematics
**********

The Kinematics Plugin
---------------------

MoveIt uses a plugin infrastructure, especially targeted towards allowing users to write their own inverse kinematics algorithms. Forward kinematics and finding jacobians is integrated within the RobotState class itself. The default inverse kinematics plugin for MoveIt is configured using the [KDL](https://github.com/orocos/orocos_kinematics_dynamics) numerical jacobian-based solver. This plugin is automatically configured by the MoveIt Setup Assistant.

> IKFast Plugin

Reviewer 3: IKFast hasn't been ported yet. Can you comment this out for now?

Author: Absolutely. Didn't realize that was the case.
